### PR TITLE
ENH: Enable the ability to compute multiple pipelines

### PR DESCRIPTION
**Pull request by lianga888**

### CODE REVIEW

File: zipline/errors.py [hunk 1]
```diff
@@ -639,6 +639,16 @@ class NoSuchPipeline(ZiplineError, KeyError):
     )
 
 
+class DuplicatePipelineName(ZiplineError):
+    """
+    Raised when a user tries to attach a pipeline with a name that already
+    exists for another attached pipeline.
+    """
+    msg = (
+        "Pipeline name '{name}' already exists for another pipeline."
```
Comment (ssanderson) [inline zipline/errors.py]: You can also use `{name!r}` here instead of having quotes around the format unit.

Comment (ssanderson) [reply zipline/errors.py]: Might be helpful to include a suggestion of how to fix the issue here, e.g. "Tried to attach a pipeline named {name}, but there was already an attached pipeline with that name. Please use a different name for this pipeline."

File: tests/pipeline/test_pipeline_algo.py [hunk 2]
```diff
@@ -344,6 +349,72 @@ def handle_data(context, data):
...
+                # Assets should appear iff they exist today and yesterday.
+                exists_today = self.exists(date, asset)
+                existed_yesterday = self.exists(date - self.trading_day, asset)
+                if exists_today and existed_yesterday:
+                    self.assertEqual(
+                        closes.loc[asset, 'close'],
+                        self.expected_close(date, asset)
+                    )
+                    self.assertEqual(
+                        volumes.loc[asset, 'close'],
```
Comment (ssanderson) [inline tests/pipeline/test_pipeline_algo.py]: I think this column name is a typo.  I'm surprised that this test currently passes.

File: tests/pipeline/test_pipeline_algo.py [hunk 3]
```diff
@@ -344,6 +349,72 @@ def handle_data(context, data):
...
+                        self.expected_close(date, asset)
```
Comment (ssanderson) [inline tests/pipeline/test_pipeline_algo.py]: Should this be `expected_volume`?

File: tests/pipeline/test_pipeline_algo.py [hunk 4]
```diff
@@ -181,6 +183,9 @@ def expected_close(self, date, asset):
             lookup = self.adj_closes
         return lookup.loc[date, asset]
 
+    def expected_volume(self, date, asset):
```
Comment (ssanderson) [inline tests/pipeline/test_pipeline_algo.py]: I would expect this function to have to account for the split in the test data used by this test.

**Comment by lianga888** [comment]
@ssanderson this PR is ready for another pass, at your convenience

File: zipline/utils/cache.py [hunk 5]
```diff
@@ -143,8 +174,14 @@ def get(self, key, dt):
         try:
             return self._cache[key].unwrap(dt)
         except Expired:
-            del self._cache[key]
-            raise KeyError(key)
+            # We can't handle safely handle the exception in this block
```
Comment (ssanderson) [inline zipline/utils/cache.py]: If we're going to re-raise an exception, then we don't need to explicitly clear the exception in the cleanup function.  The important thing is just that by the time we exit this function, the exception referencing the cached object is no longer in `sys.exc_info`.

File: zipline/utils/cache.py [hunk 6]
```diff
@@ -119,7 +148,7 @@ def __init__(self, cache=None):
         else:
             self._cache = {}
 
-    def get(self, key, dt):
+    def get(self, key, dt, cleanup=basic_cleanup):
```
Comment (ssanderson) [inline zipline/utils/cache.py]: I would probably make this a parameter to `ExpiringCache` instead of to `get`.

File: zipline/utils/cache.py [hunk 7]
```diff
@@ -86,6 +88,33 @@ def _unsafe_get_value(self):
...
+    # 1. self._pipeline_cache holds a reference.
+    # 2. The dataframe itself holds a reference via cached .iloc/.loc
+    #    accessors.
+    # 3. The traceback held in sys.exc_info includes stack frames in
+    #    which self._pipeline_cache is a local variable.
+
+    # We remove the above sources of references in reverse order:
+
+    # 3. Clear the traceback.  This is no-op in Python 3.
+    exc_clear()
```
Comment (ssanderson) [inline zipline/utils/cache.py]: I don't think this is necessary anymore.  See note below.

File: zipline/utils/cache.py [hunk 8]
```diff
@@ -86,6 +88,33 @@ def _unsafe_get_value(self):
...
+def basic_cleanup(cache, key):
+    del cache[key]
+
+
+def dataframe_gc_cleanup(cache, key):
+    # Try to deterministically garbage collect the previous result by
+    # removing any references to it. There are at least three sources
+    # of references:
+
+    # 1. self._pipeline_cache holds a reference.
```
Comment (ssanderson) [inline zipline/utils/cache.py]: This comment is referencing variables from TradingAlgorithm, which doesn't really make sense if this is a standalone function.

File: zipline/utils/cache.py [hunk 9]
```diff
@@ -86,6 +88,33 @@ def _unsafe_get_value(self):
...
+    # We remove the above sources of references in reverse order:
+
+    # 3. Clear the traceback.  This is no-op in Python 3.
+    exc_clear()
+
+    # 2. Clear the .loc/.iloc caches.
+    clear_dataframe_indexer_caches(cache[key]._unsafe_get_value())
+
+    # 1. Clear the reference to the `CachedObject`
+    del cache[key]
```
Comment (ssanderson) [inline zipline/utils/cache.py]: I don't think the cleanup function should have to be responsible for clearing the cache key.  That should happen unconditionally no matter what.

I think the right signature for the cleanup function is to just take the value wrapped in the `CachedObject`.

**Comment by ssanderson** [comment]
LGTM

**Pull request merged**